activesupport/lib/active_support/json_with_marshal_fallback.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  class JsonWithMarshalFallback
+    MARSHAL_SIGNATURE = "\x04\x08"
+
+    cattr_accessor :fallback_to_marshal_deserialization, instance_accessor: false, default: true
+    cattr_accessor :use_marshal_serialization, instance_accessor: false, default: true
+
+    class << self
+      def dump(value)
+        if self.use_marshal_serialization
+          Marshal.dump(value)
+        else
+          JSON.encode(value)
+        end
+      end
+
+      def load(value)
+        raise ::JSON::ParserError if value.start_with?(MARSHAL_SIGNATURE)
+        JSON.decode(value)
+      rescue ::JSON::ParserError
+        if self.fallback_to_marshal_deserialization
+          Marshal.load(value)
+          Rails.logger.error("JsonWithMarshalFallback: Marshal load fallback occured.")
```

Comment:
I'd suggest `warn` here, since this is something that the developer should address, but that the application can recover from without issue.
```suggestion
          Rails.logger.warn("JsonWithMarshalFallback: Marshal load fallback occured.")
```